sdks/java/harness/src/main/java/org/apache/beam/fn/harness/control/ExecutionStateSampler.java

```diff
@@ -167,9 +198,15 @@ private Void stateSampler() throws Exception {
         Thread.sleep(difference);
       } else {
         long millisSinceLastSample = currentTimeMillis - lastSampleTimeMillis;
+        Optional<String> timeoutMsg = Optional.empty();
         synchronized (activeStateTrackers) {
           for (ExecutionStateTracker activeTracker : activeStateTrackers) {
-            activeTracker.takeSample(currentTimeMillis, millisSinceLastSample);
+            timeoutMsg =
+                timeoutMsg.or(
+                    () -> activeTracker.takeSample(currentTimeMillis, millisSinceLastSample));
+            if (timeoutMsg.isPresent() && this.onTimeoutExceededCallback != null) {
```

Comment:
move this outside the synchronized block so we're not calling the callback under it.

we're only going to look at the first timeoutMsg but that is fine.